docs/adr/0008-wrapper-code.md

```diff
@@ -0,0 +1,91 @@
+
+#### UniFFI-supported interfaces
+
+One last reason for wrapper code is to present idiomatic interfaces to our Rust code -- especially for callback interfaces.
+For example, it's possible to define a UniFFI callback interface for notifications, but the FxA wrapper code uses the `NotificationCenter` on Swift which is not supported by UniFFI.
+If we wanted to remove all wrapper code we would need to commit to only using interfaces that UniFFI could support.
+
+## Decision Drivers
+
+## Considered Options
+
+* **(A) Keep using our current strategy**
+Don't invest time into removing wrapper code.
+
+* **(B) Remove all wrapper code**
+Invest time into removing wrapper code with the intention of removing all of it.
+
+* **(C) Remove most wrapper code, keep an additive wrapper layer**
+Invest time into removing wrapper code, but keep some wrapper code.
+In particular, keep a wrapper layer that adds new API surfaces rather than replaces existing ones.
+For example, we could define an `FxaEventListener` interface with UniFFI, then add a `IosEventListener` class that implemented that interface by forwarding the messages to the `NotificationCenter`.
+
+## Decision Outcome
+
+## Pros and Cons of the Options
+
+### (A) Keep using our current strategy
+
+* Good, because we can spend our time on other improvements
+* Good, because there's no chance of wasting time on implementing solutions that may not work out in practice.
+
+### (B) Remove all wrapper code
+* Good, because it simplifies our documentation strategy.
+  There is active work in UniFFI to auto-generate the bindings documentation from the Rust docstrings (https://github.com/mozilla/uniffi-rs/pull/1498, https://github.com/mozilla/uniffi-rs/pull/1493).
+  If there are no wrappers, then we could potentially use this to auto-generate a high-level documentation site and/or docstrings in the generated bindings code.
+  If there are wrappers, then this probably isn't going to work.
+  In general, wrappers mean we are have multiple public APIs which makes documentation harder.
+* Bad, because it may lead to worse documentation.
+  Hand-written documents can be better than auto-generated ones and especially since they can specifically target javadoc and swiftdoc.
+* Good, because a "vanilla" API may be easier to integrate with multiple consumer apps.
+  `NotificationCenter` might be the preferred choice for firefox-ios, but another Swift app may want to use a different system.
+  By only using UniFFI-supported types we can be fairly sure that our code will work with any system.
+
+### (C) Remove most wrapper code, keep an additive wrapper layer
+* Good, because most documentation can be auto-generated and some can be hand-written.
+  The hand-written documentation would be the language-specific parts, which probably need to be written by hand.
+* Bad because it may lead to worse documentation, for the same reasons an (B).
+* Good, because consumer apps can choose to use the wrapper interfaces or not.
```

Comment:
I think I'd be leaning towards an option described as something like "as few wrappers as possible, but no less" - ie, a perfect world has no wrappers, but this world is going to force *some* on us. Deciding "no wrappers" would be a fool's errand and not survive reality.

I guess this is closest to (C) Is there some feature uniffi is missing to enable this without the "bad" parts? It almost gets back to the "decorators" idea we had all that time ago - ie, something more like "augmentation" than "wrappers"?